fix: skaffold build fails when attempting to COPY an empty dir

kadern0 opened the pull request:
Signed-off-by: Pablo Caderno <kaderno@gmail.com>

<!-- Thank you for your contribution! -->

<!-- Include if applicable: -->
Fixes: #6810 
Related: #6796 
**Merge before/after**: _Dependent or prerequisite PRs_

**Description**
<!-- Describe your changes here. The more detail, the easier the review! -->
Refactored function WalkWorkspace to also run over directories. If directory wasn't empty, by copying the file it would create also the directory, but if the directory is empty there is some code region that couldn't be reached the way it was written.


**Follow-up Work (remove if N/A)**
The code I've added has some duplication that I'm not a huge fan of that could be easily replaced if there's an easy/better way to figure out that a directory is empty.

If this is merged, the test introduced within #6796 can be deleted as this would replace it.

briandealwis commented on pkg/skaffold/docker/dependencies.go:
```diff
@@ -283,7 +283,26 @@ func WalkWorkspace(workspace string, excludes, deps []string) (map[string]bool,
+				return err
+			}
+			var keepDir = true
+			for item := range files {
+				if strings.Contains(item, relPath) {
+					keepDir = false
+				}
+			}
```
This will be very expensive and also suffers from false positives.  Consider the following:
```
test/  <-- empty directory
testing/README.txt
src/main/test/Test.java
```

`testing/README.txt` contains `test`
`src/main/test/Main.java` contains `test`

There are a few approaches I can think of here:
1. Change the original walk to process both files and directories.  If the path is a directory, check if the directory has contents, and if empty then add the directory to `files`.  This could entail adding a `walk.IsEmpty(dir string) bool` funciton, whose implementation would use [godirwalk's NewScanner](https://github.com/karrick/godirwalk/blob/f64d91544257ab0cc1cca44f22600925a1b9ddae/examples/remove-empty-directories/main.go#L50-L66).
2. Do two passes, where the first pass collects the parent directories of files to a separate `directoriesWithChildren` map. Then walk the directories in the second pass and only add directories to `files` if they weren't in `directoriesWithChildren`.
3. Add a new `DoThen(Action, postDir Action)` action to our `walk` interface that adds a post-directory action using [`godirwalk`'s PostChildrenCallback](https://github.com/karrick/godirwalk/tree/f64d91544257ab0cc1cca44f22600925a1b9ddae#configurable-post-children-callback).  Then do the same `directoriesWithChildren` building as in Option 2 but then check empty-dirs in this post-children callback.

kadern0 replied:
Thank you very much for your comment. I agree with you about the extra cost of going over all the files for each directory (although preventing false positives should be easy to do). 

My initial idea was using a function to check if the directory is empty (so basically the first approach you described). That would avoid having to make two passes. 

I'm going to have a look at the godirwalk to see how it's being done. I'll let you know if I get stuck.

briandealwis commented on pkg/skaffold/docker/dependencies.go:
```diff
@@ -260,10 +259,17 @@ func WalkWorkspace(workspace string, excludes, deps []string) (map[string]bool,
 		keepFile := func(path string, info walk.Dirent) (bool, error) {
+			// This region was never reached
 			// Always keep root folders.
-			if info.IsDir() && path == absFrom {
-				return true, nil
+			// if info.IsDir() && path == absFrom {
+			// 	return false, nil
+			// }
+
```
Remove this code (I'm not sure what line 262 is referring to)

kadern0 replied:
I wanted to show  (for discussion) that this `if` statement always evaluates to false since the `keepfile` function was only being called for files and not directories. Also, if we add the root directory tests fail so I initially set it to false but then I realized we don't need that part at all (hence commented out and ready to be removed if there's an agreement)

briandealwis replied:
Ah makes sense 👍

briandealwis replied:
These comments need to be cleaned up.

briandealwis reviewed (commented):
Thanks for the contribution!  Two nits to address.

kadern0 commented on pkg/skaffold/util/util.go:
```diff
@@ -287,6 +287,14 @@ func IsDir(path string) bool {
 	return (err == nil || !os.IsNotExist(err)) && info.IsDir()
 }
 
+func IsEmptyDir(path string) bool {
```
Wanted to know if this approach was good first. I'll add testing.

briandealwis commented on pkg/skaffold/util/util.go:
```diff
@@ -287,6 +287,14 @@ func IsDir(path string) bool {
 }
 
+func IsEmptyDir(path string) bool {
+	files, err := os.ReadDir(path)
+	if err != nil {
+		return false
+	}
+	return len(files) == 0
```
I think it would be worth optimizing this.  The [implementation of `os.ReadDir(string)`](https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/os/dir.go;l=115-125) does the following:
```golang
func ReadDir(name string) ([]DirEntry, error) {
	f, err := Open(name)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	dirs, err := f.ReadDir(-1)
	sort.Slice(dirs, func(i, j int) bool { return dirs[i].Name() < dirs[j].Name() })
	return dirs, err
}
```

We don't care about the full list nor about sorting.  So we could just use:
```suggestion
	d, err := os.Open(name)
	if err != nil {
		return false
	}
	defer d.Close()
	ents, err := d.ReadDir(1)
	if err != nil {
		return false
	}
	return len(ents) == 0
```

kadern0 replied:
Agree. That will be more efficient for this use case.

briandealwis commented on pkg/skaffold/util/util.go:
```diff
@@ -287,6 +287,20 @@ func IsDir(path string) bool {
+		return false
+	}
+	defer d.Close()
+	ents, err := d.ReadDir(0)
```
According to the [docs](https://pkg.go.dev/os#File.ReadDir), `n ≤ 0` will cause all of the entries to be read.  Setting `n = 1` will ensure that we only read the first entry.

kadern0 replied:
You're right but setting it to 1 will also return error for empty directories so it was failing. I think the documentation is not too clear:
```
If n <= 0, ReadDir returns all the DirEntry records remaining in the directory. When it succeeds, it returns a nil error (not io.EOF). 
```

You can test it yourself really quick. This is the output when I set `n = 1`:
```
=== RUN   TestIsEmptyDir
    util_test.go:457: bool differ (-got, +want):   bool(
        - 	false,
        + 	true,
          )
--- FAIL: TestIsEmptyDir (0.00s)
```

briandealwis replied:
I'm sure I tested this but I deleted my test case.  Trying again, it fails just like you say 🤦 

Easy fix though:
```
	if _, err := d.ReadDir(1); err == io.EOF {
		return true
	}
	return false
```

kadern0 replied:
I like this one, thanks! :+1:

kadern0 commented:
@briandealwis, according to this comment on the `WalkWorkspace` function: `// Note: if you change this function, you might also want to modify `walkWorkspaceWithDestinations`.`  we might still have some work to do. 

On a quick look, this function seems to also ignore empty directories. I'm going to add a `WIP` to the PR and see what else I can find.

kadern0 commented:
Hi @briandealwis , I think I'm going to need some help here. I think this code change fixes the issue, but I cannot create an appropriate test for it. I was trying to modify the integration test but I'm struggling with the empty directory creation since you cannot add empty dirs on git (which I did not know until working on this). The error thrown on the integration tests means that the `testdir` is not being created where it was supposed to so I'm wondering how to do that properly. I'll be offline for 10 days so I won't be able to work on this until then.

SSGoku369 reviewed (commented):
ok

SSGoku369 commented:
ok

MarlonGamez commented:
hi @kadern0, sorry for disappearing on this PR for so long. Are you still interested on working on this? We can try to help out, however, we're planning on a code freeze for a bit this upcoming week in order to prepare for a large merge we're doing. If you're still interested in working on this, we can leave it open and pick it up once that merge is done

kadern0 commented:
Hi @MarlonGamez, happy to work on this after the merge, although I think I need a hand to create a proper test for this.

aaron-prindle commented:
Friendly ping @kadern0, sorry for the delay with the PR here.  Is there any change you can resubmit your changes such that we can get CI green and we can get this over the finish line?  Thanks!

kadern0 commented:
Thanks for the ping. I'm finishing my holidays, should be able to work on
it towards the end of next week.

On Tue, 26 July 2022, 20:33 Aaron Prindle, ***@***.***> wrote:

> Friendly ping @kadern0 <https://github.com/kadern0>, sorry for the delay
> with the PR here. Is there any change you can resubmit your changes such
> that we can get CI green and we can get this over the finish line? Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleContainerTools/skaffold/pull/6851#issuecomment-1195841410>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADEHQTBFKIGNYR6OU4QMN43VWAVOLANCNFSM5H3ZJACA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kadern0 commented:
@aaron-prindle, I've rebased the branch and quickly performed some tests locally. Whilst it seems to be doing what's expected, I've removed the WIP from the PR so all the tests are triggered although some further inspection would be appreciated.

aaron-prindle commented:
Seems there is currently a test failure related to the test added:
```
    util.go:246: Timed out waiting for pods [empty-dir] in namespace "skaffoldpwspk"
    --- FAIL: TestRun/copying-empty-directory (325.34s)
```

aaron-prindle reviewed (approved):
LGTM.  Thanks for your work here @kadern0 and sorry for the delays!!
